common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDBIterator.java

```diff
@@ -182,23 +189,21 @@ public boolean skip(long n) {
 
   @Override
   public synchronized void close() throws IOException {
-    db.notifyIteratorClosed(this);
     if (!closed) {
-      it.close();
-      closed = true;
-      next = null;
+      try {
+        cleanable.clean();
```

Comment:
This is changing the locking semantics. Is this intentional ?

Reply:
@mridulm Are you referring to the addition of `synchronized (rocksDB.getRocksDB())` in `cleanable.clean()` compared to before? Perhaps we can let the `close()` method still use the original logic as follows:

1.  Add a state variable for ResourceCleaner and let RocksDB/LevelDBIterator hold an instance of ResourceCleaner, like
```java
private final AtomicBoolean status = new AtomicBoolean(true);
void statusToFalse() {
   status.set(false);
}

@Override
public void run() {
    if (status.compareAndSet(true, false)) {
      ....
     }
}
```

2.  Add a method to RocksDB/LevelDBIterator to prevent ResourceCleaner from actually releasing resources, like

```java
private void cancelXXX() {
    this.resourceCleaner.statusToFalse();
    this.cleanable.clean();
  }
```

3. change the `close()` to 

```java
  public synchronized void close() throws IOException {
    db.notifyIteratorClosed(this);
    if (!closed) {
      it.close();
      closed = true;
      next = null;
      this.cancelXXX();
    }
  }
```

Reply:
I am not necessarily saying it is an incorrect change - I have not analyzed in detail.
But in context of `Cleaner` related change, I would have expected the modifications to be equivalent and not require MT safety impact :-)

Reply:
I personally prefer to keep the method of manually closing as it is through some refactoring, without locking the db.

Reply:
> I personally prefer to keep the method of manually closing as it is through some refactoring, without locking the db.

WDYT? @mridulm @beliefer @zhaomin1423 Thanks ~

Reply:
> I personally prefer to keep the method of manually closing as it is through some refactoring, without locking the db.

sorry, I don't undenstand this implementation how to do it?
```
/**
   * Closes the given iterator if the DB is still open. Trying to close a JNI LevelDB handle
   * with a closed DB can cause JVM crashes, so this ensures that situation does not happen.
   */
  void closeIterator(LevelDBIterator<?> it) throws IOException {
    notifyIteratorClosed(it);
    synchronized (this._db) {
      DB _db = this._db.get();
      if (_db != null) {
        it.close();
      }
    }
  }
```

Reply:
Sounds good to me @LuciferYang, let us keep semantics as is. If there is a change in behavior we need, we can do it after.

Reply:
@zhaomin1423 Overall, we want to make the close method of `RocksDBIterator` look similar. The `notifyIteratorClosed ` method in `RocksDB`  needs to be retained (it can be refactored for reuse by `ResourceCleaner`).

```
  public synchronized void close() throws IOException {
    db.notifyIteratorClosed(this);
    if (!closed) {
      it.close();
      closed = true;
      next = null;
      this.cancelXXX();
    }
  }
```

The purpose of `this.cancelXXX();`  is to prevent `ResourceCleaner` from executing repeatedly when manually closed.

Reply:
@zhaomin1423 Overall, we want to make the close method of `RocksDBIterator` look similar. The `notifyIteratorClosed ` method in `RocksDB`  needs to be retained (it can be refactored for reuse by `ResourceCleaner`).

```
  public synchronized void close() throws IOException {
    db.notifyIteratorClosed(this);
    if (!closed) {
      it.close();
      closed = true;
      next = null;
      this.cancelXXX();
    }
  }
```

The purpose of `this.cancelXXX();`  is to prevent `ResourceCleaner` from executing repeatedly when manually closed.

Reply:
Thanks, sorry for late because I am busy recently, I will update it.

Reply:
```
private void cancelXXX() {
    this.resourceCleaner.statusToFalse();
    this.cleanable.clean();
  }
```
is ```this.cleanable.clean() ``` redundant? it can't be executed because the status is false.

Reply:
This is to have it removed from the Cleaner's queue as quickly as possible.

Reply:
Get, thanks